### PR TITLE
Feat/hex support

### PR DESCRIPTION
**Pull request by fiveisgreen**
Hexidecimal register support added, as well as a test for Hex support.

### CODE REVIEW

File: src/CommandBasedBackendRegisterAccessor.cpp
Comment (killenb) [inline src/CommandBasedBackendRegisterAccessor.cpp:1]: Line 206 does not produce hex representation strings.

File: tests/DummyServerLib/DummyServer.cpp [hunk 1]
```diff
@@ -193,6 +193,46 @@ void DummyServer::mainLoop() {
     else if(data == "ACC? AXIS2") {
       _serialPort->send("AXIS_2=" + std::to_string(acc[1]));
     }
+
+    else if(data.find("HEX ") == 0) {
+      auto tokens = tokenise(data);
+      if(tokens.size() <= 3) {
+        _serialPort->send("12345 Syntax error: HEX needs axis and value");
+      }
```
Comment (killenb) [inline tests/DummyServerLib/DummyServer.cpp:201]: You picked an unnecessarily complicated syntax for the hex command. Just positional, fixed number arguments would be sufficient. The rest is tested already, so we can concentrate on the hex part of it.
We should also test the mixed writing (last argument must not have 0x extension)
```HEX 0x12bab 0xd0d010 acbdef```
```
HEX?
0x12bab 0xd0d010 acbdef
```

File: tests/testCommandBasedBackendUnified.cpp [hunk 2]
```diff
@@ -256,6 +256,46 @@ struct ArrayFloatSingleLine : public RegisterDescriptorBase {
 
 /**********************************************************************************************************************/
 
+struct ArrayHexMultiLine : public RegisterDescriptorBase {
+  static std::string path() { return "/myHex"; }
+  static bool isWriteable() { return true; }
```
Comment (killenb) [inline tests/testCommandBasedBackendUnified.cpp:261]: Hm, the test is writing. It is enabled, but why does it not fail?

File: tests/testCommandBasedBackendUnified.cpp [hunk 3]
```diff
@@ -256,6 +256,46 @@ struct ArrayFloatSingleLine : public RegisterDescriptorBase {
...
+    }
+    return {v};
+  }
+
+  template<typename UserType>
+  std::vector<std::vector<UserType>> getRemoteValue([[maybe_unused]] bool raw = false) {
+    assert(!raw);
+    std::vector<UserType> v(nElementsPerChannel());
+    for(uint32_t e = 0; e < nElementsPerChannel(); ++e) {
+      v[e] = userTypeToUserType<UserType>(dummyServer.hex[e]);
```
Comment (killenb) [inline tests/testCommandBasedBackendUnified.cpp:281]: Aha, that's why the test passes. It does not require the text to be a hex representation. This is just taking the integer string, interprets it as integer and returns the correct number.
 It is missing force0xHex() here.
 
 I am wondering if we should interpret the string and store the value, and then getRemoteValue() returns the stored value. In principle it does not matter where we interpret the string. However, we also need it in setRemoteValue, so we should do it in  DummyServer::setHex().

**Review by killenb — CHANGES_REQUESTED** [review]
The writing is not working. The test does not detect it due to a bug.

What is also needed it to test that we can define writing with and without the leading ```0x```.

File: src/CommandBasedBackendRegisterAccessor.cpp [hunk 4]
```diff
@@ -162,13 +166,20 @@ namespace ChimeraTK {
...
-        buffer_2D[0][i] = userTypeToUserType<UserType, std::string>(valueMatch[i + 1]);
+      if(_registerInfo.internalType != CommandBasedBackendRegisterInfo::InternalType::HEX) {
+        for(size_t i = 0; i < _numberOfElements; ++i) {
+          buffer_2D[0][i] = userTypeToUserType<UserType, std::string>(valueMatch[i + 1]);
+        }
+      }
+      else {
+        for(size_t i = 0; i < _numberOfElements; ++i) {
+          buffer_2D[0][i] = userTypeToUserType<UserType, std::string>(force0xHex(valueMatch.str(i + 1)));
+        }
```
[the remark this answers is not in the document]
Comment (fiveisgreen) [reply src/CommandBasedBackendRegisterAccessor.cpp:182]: While this is nice, it creates a hazard where 0x is omitted in the config file, resulting in userTypeToUserType receiving "0x0x1234". So, there ought to then be a check that that 0x is absent, else throw an error.

File: include/stringUtils.h [hunk 5]
```diff
@@ -43,3 +43,16 @@
...
+ * This is useful for userTypeToUserType to recognize the number as hex and interpret it correctly.
+ *
+ * Behavior: 
+ * "0xABC" -> "0xABC"
+ * "0XABC" -> "0XABC"
+ * "ABC" -> "0xABC"
+ * "NAN" -> "0xNAN".
+ * "" -> "0x"
+ */
+[[nodiscard]] std::string force0xHex(const std::string& str) noexcept;
```
[the remark this answers is not in the document]
Comment (fiveisgreen) [reply include/stringUtils.h:58]: Ok, we can do that. We still need that logic in the DummyServer, but can be moved there. Removing it also creates the issue

File: src/CommandBasedBackend.cpp
[the remark this answers is not in the document]
Comment (fiveisgreen) [reply src/CommandBasedBackend.cpp:1]: The correction to the comment needs a home. Should it go directly into the master, or some new patch branch?

**Review by killenb — APPROVED** [review]
Approved after applying the suggestions.

Before merging:
* Test that all tests are passing (coding style test was again failing!)
* Squash everything into one commit (onto 6989ba9)
* Rebase onto master

**Pull request merged**